File: backend/src/main/java/com/example/medical/ai/AIService.java
package com.example.medical.ai;

import com.example.medical.exception.ExternalServiceException;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Service
public class AIService {

    private final WebClient webClient;
    private final AIRequestLogRepository logRepository;

    public AIService(@Value("${ai.service.base-url:http://localhost:8000}") String baseUrl,
                     AIRequestLogRepository logRepository) {
        this.webClient = WebClient.builder()
            .baseUrl(baseUrl)
            .build();
        this.logRepository = logRepository;
    }

    public Map<String, Object> analyzeReport(UUID reportId) {
        return callEndpoint("/ocr/analyze-report", Map.of("report_id", reportId.toString()));
    }

    public Map<String, Object> chat(UUID patientId, String question) {
        Map<String, Object> aiResponse = callEndpoint("/chat", Map.of(
            "patient_id", patientId != null ? patientId.toString() : "",
            "question", question
        ));
        
        // Map AI service response to frontend expected format
        // AI service returns: { "answer": "...", "citations": [...], "safety_banner": "...", "model_version": "..." }
        // Frontend expects: { "response": "..." }
        return Map.of(
            "response", aiResponse.getOrDefault("answer", "I'm sorry, I couldn't process that request."),
            "citations", aiResponse.getOrDefault("citations", List.of()),
            "safetyBanner", aiResponse.getOrDefault("safety_banner", "")
        );
    }

    public Map<String, Object> predictSymptoms(List<String> symptoms, Map<String, Object> demographics, Map<String, Object> vitals) {
        return callEndpoint("/predict/symptoms", Map.of(
            "symptoms", symptoms,
            "demographics", demographics != null ? demographics : Map.of(),
            "vitals", vitals != null ? vitals : Map.of()
        ));
    }

    private Map<String, Object> callEndpoint(String path, Map<String, Object> payload) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> response = webClient.post()
                .uri(path)
                .bodyValue(payload)
                .retrieve()
                .bodyToMono(Map.class)
                .timeout(Duration.ofSeconds(10))
                .block();
            AIRequestLog log = new AIRequestLog();
            log.setRequestType(path);
            log.setStatus("SUCCESS");
            String summary = response != null ? response.toString() : null;
            log.setResponseSummary(summary != null && summary.length() > 250 ? summary.substring(0, 250) : summary);
            logRepository.save(log);
            return response;
        } catch (Exception ex) {
            AIRequestLog log = new AIRequestLog();
            log.setRequestType(path);
            log.setStatus("ERROR");
            String errorMsg = ex.getMessage();
            log.setResponseSummary(errorMsg != null && errorMsg.length() > 250 ? errorMsg.substring(0, 250) : errorMsg);
            logRepository.save(log);
            throw new ExternalServiceException("AI service call failed", ex);
        }
    }
}

File: frontend/src/services/aiService.js
import apiClient from './api'

const aiService = {
  // Analyze medical report with AI
  analyzeReport: async (reportId) => {
    const response = await apiClient.post(`/ai/analyze-report/${reportId}`)
    return response.data
  },

  // Check symptoms with AI
  checkSymptoms: async (symptoms) => {
    const response = await apiClient.post('/ai/symptom-check', {
      symptoms: symptoms,
    })
    return response.data
  },

  // Get AI chat response
  sendChatMessage: async (message) => {
    const response = await apiClient.post('/ai/chat', {
      message: message,
    })
    return response.data
  },

  // OCR text extraction from report
  extractReportText: async (reportId) => {
    const response = await apiClient.post(`/ai/ocr/${reportId}`)
    return response.data
  },

  // Generate medical insights
  generateInsights: async (patientId) => {
    const response = await apiClient.post(`/ai/insights/${patientId}`)
    return response.data
  },

  // Predict health risks
  predictRisks: async (patientId) => {
    const response = await apiClient.post(`/ai/predict-risks/${patientId}`)
    return response.data
  },
}

export default aiService
